llvm/lib/Target/SPIRV/SPIRVCallLowering.cpp

```diff
@@ -194,23 +194,39 @@ getKernelArgTypeQual(const Function &KernelFunction, unsigned ArgIdx) {
   return {};
 }
 
-static Type *getArgType(const Function &F, unsigned ArgIdx) {
+static SPIRVType *getArgSPIRVType(const Function &F, unsigned ArgIdx,
+                                  SPIRVGlobalRegistry *GR,
+                                  MachineIRBuilder &MIRBuilder) {
+  // Read argument's access qualifier from metadata or default
+  SPIRV::AccessQualifier::AccessQualifier ArgAccessQual =
+      getArgAccessQual(F, ArgIdx);
+
   Type *OriginalArgType = getOriginalFunctionType(F)->getParamType(ArgIdx);
-  if (F.getCallingConv() != CallingConv::SPIR_KERNEL ||
-      isSpecialOpaqueType(OriginalArgType))
-    return OriginalArgType;
+
+  // In case of non-kernel SPIR-V function, use the original IR type.
+  if (F.getCallingConv() != CallingConv::SPIR_KERNEL)
+    return GR->getOrCreateSPIRVType(OriginalArgType, MIRBuilder, ArgAccessQual);
+
+  // Use original type if it is already TargetExtType of a builtin type.
+  if (isSpecialOpaqueType(OriginalArgType))
+    return GR->getOrCreateSPIRVType(OriginalArgType, MIRBuilder, ArgAccessQual);
 
   MDString *MDKernelArgType =
       getKernelArgAttribute(F, ArgIdx, "kernel_arg_type");
-  if (!MDKernelArgType || !MDKernelArgType->getString().endswith("_t"))
-    return OriginalArgType;
-
-  std::string KernelArgTypeStr = "opencl." + MDKernelArgType->getString().str();
-  Type *ExistingOpaqueType =
-      StructType::getTypeByName(F.getContext(), KernelArgTypeStr);
-  return ExistingOpaqueType
-             ? ExistingOpaqueType
-             : StructType::create(F.getContext(), KernelArgTypeStr);
+  if (!MDKernelArgType || (MDKernelArgType->getString().ends_with("*") &&
```

Comment:
My main question is why is the pointer type handled by parsing the typename and not using the IR types?
For HLSL specific types, they do the following:

```
%"class.hlsl::RWBuffer" = type { ptr }
```

Then, they can use this type as a normal IR type.
And then, the type can be annotated using LLVM's addresspace notation, and other constructs. For read-only buffers, they mark it as constant for ex.
It seems OpenCL does it using several metadatas like "kernel_arg_addr_space" and "kernel_arg_access_qual".
Is that just for historical reasons, or is there a real benefit in doing this over what HLSL is doing?

Reply:
@Keenuts I believe this is for historical reasons. These are the metadata we get from the Khronos SPIR-V Translator and Clang. The situation is even more complicated since older versions of the SPIR-V Translator emit these metadata in other format (not attached to functions) and we need to handle both formats.

Reply:
Right, I see, so we would now have 4 sources:
 - old translator, with the old format
 - new translator, new format
 - opencl frontend
 - hlsl frontend
 
Do you think it would be valuable to add 3 normalization passes before lowering?
Each pass would convert from 1 of those format to the one we deem canonical, so the backend then only has to handle one type, and if we retire one format, we can do it by removing the pass?